examples/bind-shopping.amp.html

```diff
@@ -0,0 +1,454 @@
+    }
+    .thumbnails amp-img[selected] {
+      border: solid 2px gray;
+    }
+    .thumbnails amp-img[out-of-stock] {
+      background-color: red;
+    }
+    amp-accordion {
+      margin-top: 10px;
+    }
+    .sizes button {
+      display: block;
+      width: 100%;
+      border: solid 1px gray;
+      padding: 5px;
+      background-color: #eee;
+    }
+    .change-size {
+      display: block;
+      width: 200px;
+      height: 30px;
+      padding: 5px;
+      background-color: #666;
+      color: white;
+    }
+    button[hidden] {
+      display: none;
+    }
+    p[hidden] {
+      display:none;
+    }
+  </style>
+</head>
+
+<body>
+  <h1 class="header">AMPWEAR</h1>
+  <div class="container">
+    <h2>MEN'S COTTON SHORT SLEEVE T-SHIRT</h2>
+
+    <amp-bind-dataset layout=nodisplay>
+      <script type="application/json">
+        {
+          "items": [
+            {
+              "id": "0",
+              "name": "09 BLACK",
+              "image": "./shirts/black.jpeg",
+              "sizes": ["S", "L"]
+            },
+            {
+              "id": "1",
+              "name": "34 BROWN",
+              "image": "./shirts/brown.jpeg",
+              "sizes": ["M", "L"]
+            },
+            {
+              "id": "2",
+              "name": "03 GRAY",
+              "image": "./shirts/gray.jpeg",
+              "sizes": ["S"]
+            },
+            {
+              "id": "3",
+              "name": "69 NAVY",
+              "image": "./shirts/navy.jpeg",
+              "sizes": ["S", "L"]
+            },
+            {
+              "id": "4",
+              "name": "19 WINE",
+              "image": "./shirts/wine.jpeg",
+              "sizes": []
+            },
+            {
+              "id": "5",
+              "name": "61 BLUE",
+              "image": "./shirts/blue.jpeg",
+              "sizes": ["L"]
+            },
+            {
+              "id": "6",
+              "name": "59 DARK GREEN",
+              "image": "./shirts/dark-green.jpeg",
+              "sizes": ["S"]
+            },
+            {
+              "id": "7",
+              "name": "02 LIGHT GRAY",
+              "image": "./shirts/light-gray.jpeg",
+              "sizes": ["S", "M", "L"]
+            },
+            {
+              "id": "8",
+              "name": "00 WHITE",
+              "image": "./shirts/white.jpeg",
+              "sizes": []
+            }
+          ]
+        }
+      </script>
+    </amp-bind-dataset>
+
+    <!--
+    <amp-carousel slide=0 id="carousel-1" width=1 height=1 layout=responsive type=slides controls
+        bind-attr="{slide: selected}"
+        on="slidechange:setState({selected: ampEventData})">
+    -->
+    <amp-carousel slide=0 id="carousel-1" width=1 height=1 layout=responsive type=slides controls>
+      <amp-bind-attribute layout=nodisplay
+          attr="slide"
+          value="{{selected}}">
+      </amp-bind-attribute>
+      <amp-bind-event layout=nodisplay
```

Comment:
Hmm, good point. That markup could look like this:
```
<amp-carousel slide=0 id="carousel-1">
</amp-carousel>

<amp-img id="img-1"
    on="tap:#carousel-1.setSlide(0)"
    bind-attr="{selected: #carousel-1.slide === 0}">
</amp-img>
```
This would save a few LOC and an indirection. However, now the "current selection" state lives in `#carousel-1` and all references of `selected` will need to be replaced by `#carousel-1.slide`. If we instead keep `selected` var, then the state lives in two places. So I think this may be locally convenient but globally inconvenient.

And then imagine a third element being bound to `#img-1`, and then `#carousel-1` can be bound to it creating a cycle, etc. If developers aren't careful, this can create a sort of "bind spaghetti" where application state is difficult to keep track of.

My understanding is that frameworks like Flux, Redux, etc. advocate making a similar tradeoff  ("unidirectional data flow") to avoid this class of problem.

[Redux says](http://redux.js.org/docs/introduction/Motivation.html):
> UI state is also increasing in complexity, as we need to manage the active route, the selected tab, whether to show a spinner or not, should pagination controls be displayed, and so on.

> Managing this ever-changing state is hard. If a model can update another model, then a view can update a model, which updates another model, and this, in turn, might cause another view to update. At some point, you no longer understand what happens in your app as you have lost control over the when, why, and how of its state.

So I think forcing all bindings to update local state first:
- Avoids "bind spaghetti", cycles, etc. by developers
- Makes applying performance constraints to eval/update easier and more predictable
- Is familiar to React developers
- Is more straightforward to implement

It's not a perfect comparison since amp-bind is not meant to manage a web app's entire state, unlike these frameworks. At least, I think not having "bind-to-element-state" may be sufficient for V1. I'd defer to your judgment and experience though. :)